src/psyclone/psyir/nodes/routine.py

```diff
@@ -160,6 +187,41 @@ def node_str(self, colour=True):
         '''
         return self.coloured_name(colour) + "[name:'" + self.name + "']"
 
+    @property
+    def _parent(self):
+        return self._parent_node
+
+    @_parent.setter
+    def _parent(self, parent):
```

Comment:
While I am not asked for a review, while studying the problem of `call_tree_utils`, I noticed this code. I think this functionality should be implemented as an explicit function with a descriptive name. I saw the code:
```
node._parent = None
```
There is no way to know that this will actually modify a symbol table (and really annoying while debugging). While I am aware that my attitude towards properties is different from the rest of the PSyclone team, there is imho just too much happening as a side affect of a simple assignment.

Reply:
I see your point here with the side effects, but the challenge is the `Routine` is in charge of controlling where its `RoutineSymbol` lives now already.
This `_parent` set (or rather the ChildrenList setting up parent links) also has a number of side effects that are somewhat not obvious anyway depending on the tree I think since the `update_signal` happens anyway I think from the ChildrenList.

Reply:
I think `_parent` should almost always only be set from functions (e.g. `detach()`).

Reply:
What about overwriting `detach` in `Routine` instead? That is much easier to see (e.g. if you debug, immediately when single stepping into `detach` you would notice that something special is happening).

From my point of view, the typical pattern is that `_x` is the attribute, and then `x` a property 'function'. This breaks this convention.

Reply:
I think unfortunately there is an issue because `_parent` is almost always actually set from `ChildrenList.set_parent_link` or other similar functions (e.g. when using addchild). This also meant I had to override `_parent`.

The specific choice of using `_parent` as the setter here is to keep it marked as protected as setting the parent of a Node is not something intended to be widely accessed, and this is required to match the rest of the code. Its a slightly awkward implementation, but I don't think there's another way around it.

Reply:
@hiker  I share your concerns about properties doing too much magic under-the-hood. But I think the deciding factor is if what they do is unexpected, or the opposite - not doing it would be more unexpected. Also, it's good that the public API, including properties, are atomic operations (wrt not leaving the tree in an inconsistent state, with a mandatory fix expected from the user).

In the case of _parent, it is part of a double-linked relationship with children. And before we had it automatically updated it was broken after some transformations because the responsability was on the transformation writter to "fix" the doble-linked list. This has not been problem anymore since its manages by the node itself.

This PR tries to solve a similar problem in that a Rountine must always have its implementation (node) as well as its declaration (parent scope routinesymbol). But we also inadvertedly break this some times as currently users are responsible to know that if they remove, rename, replace a routine, they must update the associated routinesymbol in the symboltable, maybe restore the "own_routine_symbol" tag, not leave the routinesymbols hanging, maybe update interfaces, have a way for the detached standalone routine to find its own symbol... (and the issues that @LonelyCat124  found are because we don't even remember to do/review all this things ourselfs) 

This PR started as an experiment to see if we can get rid of the finicky "own_routine_symbol" but also to see if we can make attaching/detaching routines be more atomic in that they don't levave the tree inconsistent forcing the user to remember to fix the symbols/references. Doing it in "detach" is not sufficient because any list operation in children (e.g. `container.children = [routine1]` could remove a previous existing routine). (Unfortunately) when debugging at least you would need to know that you need to step into this list operation, but I take your point that we should put the effects in descriptive methods.